covalent_indexer.go

```diff
@@ -31,91 +39,196 @@ func NewCovalentDataIndexer(processor DataHandler, server *http.Server) (*covale
 	if server == nil {
 		return nil, ErrNilHTTPServer
 	}
-
 	ci := &covalentIndexer{
 		processor: processor,
 		server:    server,
 	}
+	ci.newConnectionWSR = make(chan struct{})
+	ci.newConnectionWSS = make(chan struct{})
 
 	go ci.start()
 
 	return ci, nil
 }
 
-func (c *covalentIndexer) SetWSSender(wss *ws.WSSender) {
-	c.Lock()
-	defer c.Unlock()
-	if c.wss != nil {
-		_ = c.wss.Conn.Close()
+func (ci *covalentIndexer) SetWSSender(wss process.WSConn) {
+	ci.mutWSS.Lock()
+	if ci.wss != nil {
+		err := ci.wss.Close()
+		log.LogIfError(err)
+	}
+	ci.wss = wss
+	ci.mutWSS.Unlock()
+
+	ci.newConnectionWSS <- struct{}{}
+}
+
+func (ci *covalentIndexer) SetWSReceiver(wsr process.WSConn) {
+	ci.mutWSR.Lock()
+	if ci.wsr != nil {
+		err := ci.wsr.Close()
+		log.LogIfError(err)
 	}
-	c.wss = wss
+	ci.wsr = wsr
+	ci.mutWSR.Unlock()
+
+	ci.newConnectionWSR <- struct{}{}
+}
+
+func (ci *covalentIndexer) getWSS() process.WSConn {
+	ci.mutWSS.RLock()
+	wss := ci.wss
```

Comment:
this is not a copy. you just pass the pointer

Reply:
is the same thing with `return ci.wss`

Reply:
Yes, this was intended. Changed the code  to highlight that I don't want to copy the value. See comment in the previous PR